RZDataBinding/RZDBCoalesce.m

```diff
@@ -0,0 +1,317 @@
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+#import <objc/runtime.h>
+#import <objc/message.h>
+
+#import "RZDBCoalesce.h"
+
+#import "NSObject+RZDataBinding.h"
+
+static NSString* const kRZDBCoalesceStorageKey = @"RZDBCoalesce";
+
+#pragma mark - RZDBNotification interface
+
+@interface RZDBNotification : NSObject
+
+@property (weak, nonatomic) id target;
+@property (assign, nonatomic) SEL action;
+
+@property (strong, nonatomic) NSMutableDictionary *changeDict;
+
++ (instancetype)notificationWithInvocation:(NSInvocation *)invocation;
+
+- (void)send;
+
+@end
+
+#pragma mark - RZDBCoalesceProxy interface
+
+@interface RZDBCoalesceProxy : NSProxy
+
+@property (assign, nonatomic) __unsafe_unretained id representedObject;
+
++ (instancetype)proxyForObject:(id)object;
+
+@end
+
+#pragma mark - RZDBCoalesce private interface
+
+@interface RZDBCoalesce ()
+
+@property (assign, nonatomic) NSInteger beginCount;
+@property (strong, nonatomic, readonly) NSMutableArray *notifications;
+
+@end
+
+#pragma mark - RZDBCoalesce implementation
+
+@implementation RZDBCoalesce
+
++ (void)begin
+{
+    RZDBCoalesce *current = [self currentCoalesce];
+
+    if ( current == nil ) {
+        [self setCurrentCoalesce:[[self alloc] init]];
+    }
+    else {
+        current.beginCount++;
+    }
+}
+
++ (void)commit
+{
+    RZDBCoalesce *current = [self currentCoalesce];
+    current.beginCount--;
+
+    if ( current.beginCount == 0 ) {
+        [self setCurrentCoalesce:nil];
+        [current.notifications enumerateObjectsUsingBlock:^(RZDBNotification *notification, NSUInteger idx, BOOL *stop) {
+            [notification send];
+        }];
+    }
+}
+
++ (void)coalesceBlock:(void (^)())coalesceBlock
+{
+    NSParameterAssert(coalesceBlock);
+
+    [RZDBCoalesce begin];
+    coalesceBlock();
+    [RZDBCoalesce commit];
+}
+
++ (void)setAutoCoalesceByRunloopOnMainThread:(BOOL)autoCoalesce
+{
+    objc_setAssociatedObject(self, @selector(autoCoalesceByRunloopOnMainThread), @(autoCoalesce), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+}
+
+#pragma mark - private methods
+
++ (BOOL)autoCoalesceByRunloopOnMainThread
+{
+    return [objc_getAssociatedObject(self, _cmd) boolValue];
+}
```

Comment:
Lets leave it out till we find a good usage for it. Easier to add something in than take it out